### PR TITLE
expose all logging options through values.yaml

### PR DESCRIPTION
**Pull request by paynejacob**
https://github.com/rancher/rancher/issues/35215

# Problem
Rancher logging creates the logging resources but this requires us to expose a lot of configuration in the values.yaml.  This pr exposes a single object that can be overlayed over the logging object and allow users to configure any fields not configured by rancher logging.

# Testing
In testing this I ensured the right precedence was being applied when merging all the templates.  I also tested the following cluster types:
- aks
- rke2
- rke
- k3s (systemd)
- k3s (openrc)

### CODE REVIEW

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+####### {{$generic}}
+
+{{/* the sources are merge right to left meaning tpl is the highest prcedence and values is the lowest */}}
+{{- toYaml (merge $tpl $values $generic) -}}
+{{- end -}}
+
+{{- define "logging-operator.logging" -}}
+{{- include "logging-operator.util.merge.logging" (append . "logging-operator.logging.tpl") -}}
+{{- end -}}
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml:118]: I understand this, but we will definitely need a `rancher/docs` issue for this. Those unfamiliar with templating engines, but using `helm` may get lost.

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_helpers.tpl [hunk 2]
```diff
@@ -154,3 +154,26 @@ Set kubernetes log options if they are configured
...
 {{- end -}}
+
+{{/*Fluent Bit Image Repository */}}
+{{- define "logging-operator.fluentbitImageRepository" -}}
+{{- if .Values.debug -}}
+{{ template "system_default_registry" . }}{{ .Values.images.fluentbit_debug.repository }}
+{{- else -}}
+{{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
+{{- end -}}
+{{- end -}}
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_helpers.tpl:165]: Isn't the debug repository the same?

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/aks/logging.yaml [hunk 3]
```diff
@@ -1,104 +1,18 @@
...
-  controlNamespace: {{ .Release.Namespace }}
   fluentbit:
-    image:
-      repository: {{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
-      tag: {{ .Values.images.fluentbit.tag }}
+    disableKubernetesFilter: true
+    extraVolumeMounts:
+      - source: {{ $logPath }}
+        destination: {{ $logPath }}
+        readOnly: true
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/aks/logging.yaml:11]: This looks new vs. the generic template. What are these used for?

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/eks/logging.yaml [hunk 4]
```diff
@@ -1,105 +1,19 @@
...
-  controlNamespace: {{ .Release.Namespace }}
   fluentbit:
-    image:
-      repository: {{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
-      tag: {{ .Values.images.fluentbit.tag }}
+    disableKubernetesFilter: true
+    extraVolumeMounts:
+      - source: {{ $logPath }}
+        destination: {{ $logPath }}
+        readOnly: true
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/eks/logging.yaml:11]: same as above

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/gke/logging.yaml [hunk 5]
```diff
@@ -1,104 +1,18 @@
...
-  controlNamespace: {{ .Release.Namespace }}
   fluentbit:
-    image:
-      repository: {{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
-      tag: {{ .Values.images.fluentbit.tag }}
+    disableKubernetesFilter: true
+    extraVolumeMounts:
+      - source: {{ $logPath }}
+        destination: {{ $logPath }}
+        readOnly: true
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/gke/logging.yaml:11]: same as above

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/k3s/logging-k3s-openrc.yaml [hunk 6]
```diff
@@ -1,109 +1,19 @@
...
-  controlNamespace: {{ .Release.Namespace }}
   fluentbit:
-    image:
-      repository: {{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
-      tag: {{ .Values.images.fluentbit.tag }}
+    disableKubernetesFilter: true
+    extraVolumeMounts:
+      - source: {{ $logPath }}
+        destination: {{ $logPath }}
+        readOnly: true
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/k3s/logging-k3s-openrc.yaml:11]: same as above

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml [hunk 7]
```diff
@@ -1,113 +1,24 @@
...
   {{- end }}
   fluentbit:
     disableKubernetesFilter: true
-    image:
-      repository: {{ template "system_default_registry" . }}{{ .Values.images.fluentbit.repository }}
-      tag: {{ .Values.images.fluentbit.tag }}
+    extraVolumeMounts:
+      - source: {{ template "kubeAuditPathPrefix" . }}
+        destination: "/kube-audit-logs"
+        readOnly: true
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml:13]: same as above

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml [hunk 8]
```diff
@@ -1,113 +1,24 @@
...
-    {{- end }}
-    {{- if .Values.global.psp.enabled }}
-    security:
-      podSecurityPolicyCreate: true
-      roleBasedAccessControlCreate: true
-    {{- end }}
-    {{- with (default .Values.tolerations .Values.fluentd.tolerations) }}
+    {{- with (concat (.Values.tolerations) (.Values.fluentbit.tolerations) (.Values.additionalLoggingSources.kubeAudit.fluentbit.tolerations)) }}
     tolerations: {{- toYaml . | nindent 6 }}
     {{- end }}
```
Comment (nickgerace) [inline charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml:20]: Why do these tolerations merge differently than the cloud provider and k3s CRs?

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_helpers.tpl (hunk 2 above)
Comment (paynejacob) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_helpers.tpl:165]: it is but it doesn't have to be

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml (hunk 8 above)
Comment (paynejacob) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/kube-audit/logging.yaml:20]: Not sure if there was a specific reason originally but in order to not introduce any breaking changes I kept it the same.

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/aks/logging.yaml (hunk 3 above)
Comment (paynejacob) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/loggings/aks/logging.yaml:11]: we were not properly mounting files before so these loggings were never working.

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml (hunk 1 above)
Comment (paynejacob) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml:118]: Docs for future rancher developers or docs for users?  I am confused what we need to document?

Comment (nickgerace) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml:118]: For users

**Review by nickgerace — APPROVED** [review]

File: charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml (hunk 1 above)
Comment (paynejacob) [reply charts/rancher-logging/rancher-logging/100.0.1+up3.15.0/templates/_generic_logging.yaml:118]: I added section in the values about how to use it.  I think we could add a mention in the docs page to checkout that value if you want to do more complex things but we should discourage it's use.

**Review by rmweir — APPROVED** [review]
LGTM

**Pull request merged**